### PR TITLE
Make manifest_file work as either file.pp or a directory

### LINKED ISSUE #2373: Puppet masterless provisioner `manifest_file` directory support has many interconnected issues
:warning: The following is kind of a brain dump after wrestling with something for hours through the night on a couple of occasions, haha. Sorry if it's at all unclear.

----------------------------

So I've personally been waiting for **v0.8.0** for quite a bit now for really just one basic feature, supporting provisioning by manifest directory instead of pointing to a main file (#1446 / #1771). To work around this lack of behavior I previously had to actually copy and paste the internal provisioning execution template string and slightly modify it to allow for directories. It was gross, haha, but it worked.

[Excited for the new release][twitter-trevorsuarez-mitchellh-thanks], I installed v0.8.0, removed the modified `execute_command` string from my config, changed the `manifest_file` to a directory instead of a dummy "run_me_first" file with imports, and ran Packer. Unfortunately, I was met with errors at provisioning runtime. :/

After spending about 4-6 hours wrestling with the issue, I think I've discovered the issues that are causing the failure of this new feature/fix. Interestingly enough, I don't think its just one issue that's causing the failure.

**So where the error keeps arising is actually when the provisioner uploads the manifests.**

See, the first sign of issue is a _WARNING_ that's emitted during the upload:
> WARNING: manifest_file should be a file. Use manifest_dir for directories

This looks to have originated from #2258 (commit 8990c38d5e53f4148fbee669a5ca1703d701abe2), where for some reason a warning was put in place to denote a "deprecation" of setting the `manifest_file` configuration attribute as a directory. Now, that makes sense in that there's also a `manifest_dir` attribute... except when I tried to remove the `manifest_file` attribute and leave the `manifest_dir` attribute as the actual directory instead, I received an error:

> Template validation failed. Errors are shown below.
>
> Errors validating build 'amazon-ebs'. 1 error(s) occurred:
> 
> * A manifest_file must be specified.

So while it seemed as if using a directory path for the `manifest_file` attribute was deprecated and `manifest_dir` was intended as its replacement, it doesn't actually seem like its possible to replace without hitting the validation error. :/

Hmm, ok, so I tried to be a tiny bit tricky instead to see if I could bypass the validation by setting the `manifest_dir` to the directory correctly and setting the `manifest_file` attribute to the value `.`. I figured this would attempt to just run the directory specified in `manifest_dir` and everything would work... but I was met with another error when the provisioner attempted to upload the manifests:

> Build 'amazon-ebs' errored: Error uploading manifests: scp: /tmp/packer-provisioning/puppet/manifests/.: Is a directory

Yea, so while it looks like #1771 allowed for the `manifest_file` attribute to be set as a directory path, it looks like the `scp` communication tool either had a regression within the months time since that fix or never quite worked, as it won't actually upload the directory specified. Yea, damn.

Ok, **so I tried one last thing**. I figured that the last error maybe had something to do with the current directory workaround I was using to bypass the config validation issue, so I tried to set the `manifest_file` to the same value as the `manifest_dir`, very similarly to how I originally had it set in v0.7.5 except without the trailing file name (so `/path/` instead of `/path/file.pp`)... but that errored out at the same exact `scp` location. Except this time the error was even more strange:

> Build 'amazon-ebs' errored: Error uploading manifests: scp: /tmp/packer-provisioning/puppet/manifests/provisioning/puppet/manifests: No such file or directory

After a second of looking at it, I laughed. Do you see what happened there? It appended the `manifest_file` path onto the staging directory, so I ended up with a redundant and non-existent path. This looks to be because of another change made in #2258 (commit 742e5568363eecef3c00f602611ce6ad2e9c10a6) that made it so that the path wasn't "basenamed" if the `manifest_file` attribute value was a directory instead of a file. Now, while that makes sense, it also means that the `manifest_file` can't be set to a path value that contains any children, or else they'll be incorrectly appended to the remote path. Yea, damn.

So yea, this has been quite the adventure, haha. I love this tool/product, so let me know if there's anything I can do to clarify any of this or help in any way. Thanks!

[twitter-trevorsuarez-mitchellh-thanks]: https://twitter.com/trevorsuarez/status/613796428750716928

### PR DESCRIPTION
**Pull request by cbednarski**
- Includes documentation to clarify how you should configure packer for puppet with multiple manifests
- Fixes #2373

### CODE REVIEW

**Comment by mitchellh** [comment]
LGTM!

**Comment by Rican7** [comment]
Looks great. Smart documentation additions too. :+1:

File: website/source/docs/provisioners/puppet-masterless.html.markdown [hunk 1]
```diff
@@ -64,6 +65,11 @@ Optional parameters:
   the `manifest_file`. It is a separate directory that will be set as
   the "manifestdir" setting on Puppet.
 
+  ~> `manifest_dir` is passed to `puppet apply` as the `--manifestdir` option.
+     This option was deprecated in puppet 3.6, and is slated to be removed in
+     puppet 4.0. If you have multiple manifests you should simply use
```
Comment (mootpt) [inline website/source/docs/provisioners/puppet-masterless.html.markdown]: Is removed in Puppet 4.X: https://docs.puppetlabs.com/references/4.stable/configuration.html#manifest

File: website/source/docs/provisioners/puppet-masterless.html.markdown [hunk 2]
```diff
@@ -40,9 +40,10 @@ The reference of available configuration options is listed below.
 
 Required parameters:
 
-* `manifest_file` (string) - The manifest file for Puppet to use in order
-  to compile and run a catalog. This file must exist on your local system
-  and will be uploaded to the remote machine.
+* `manifest_file` (string) - This is either a path to a puppet manifest (`.pp`
+  file) _or_ a directory containing multiple manifests that puppet will apply.
```
Comment (mootpt) [inline website/source/docs/provisioners/puppet-masterless.html.markdown]: Is it worth noting the evaluation order, when there are multiple manifests? (i.e evaluated in alphabetical order)

**Comment by mootpt** [comment]
Looks good.

File: website/source/docs/provisioners/puppet-masterless.html.markdown (hunk 2 above)
Comment (Rican7) [inline website/source/docs/provisioners/puppet-masterless.html.markdown]: That's an internal puppet implementation detail, and could theoretically change at Puppet's discretion. I don't think that should really be Packer's responsibility to document.

Comment (mootpt) [inline website/source/docs/provisioners/puppet-masterless.html.markdown]: True. No disagreement here.

Comment (cbednarski) [inline website/source/docs/provisioners/puppet-masterless.html.markdown]: I can link to https://docs.puppetlabs.com/puppet/latest/reference/dirs_manifest.html#directory-behavior-vs-single-file, which clarifies the behavior.

File: website/source/docs/provisioners/puppet-masterless.html.markdown (hunk 1 above)
Comment (cbednarski) [inline website/source/docs/provisioners/puppet-masterless.html.markdown]: Thanks! Was confused about that.

**Pull request merged**